### 5. Algorithmic Question

First we will consider the Brute-force algorithm which go over all the possible scores and return the maximal one.    
This is clearly not the optimal solution because in this case we have running time of order O(n!)- which is higher than exponential !!  

In [23]:
#brute force solution

def exams_score(score, exams):
    if len(exams) == 1:
        return exams[0]
    else:
        return max([exams_score(element, [x + (score - element) for x in exams if x != element]) for element in exams])

In [31]:
score = int(input("Enter initial personal score: "))
arr = list(map(int, input("Enter exams marks: ").split()))
exams_score(score, arr)

Enter initial personal score: 25
Enter exams marks: 18 24 21 32 27


44

### Optimal Implementation
The following solution is the **optimal one:**

In [20]:
#optimized implementation
import math

def max_score(score, arr):
    if len(arr) == 0:
        return score
    index = math.ceil(len(arr) / 2) - 1
    next_score = arr[index]
    return max_score_calc(next_score, [x + (score - next_score) for x in arr if x != next_score])

def max_score_calc(score, arr):
    return max_score(score, sorted(arr))

In [21]:
score = int(input("Enter initial personal score: "))
arr = list(map(int, input("Enter exams marks: ").split()))
max_score_calc(score, arr)

Enter initial personal score: 30
Enter exams marks: 13 27 41 59 28 33 39 19 52 48 55 79


205

### Proof of Correctness

**Base Case**: The base case is when the array `arr` is empty, n = 0. In this case, the function max_score returns the initial score `score`, which is correct because there are no more exams to take.

**Inductive Hypothesis**: Assume that the algorithm works correctly for an array of size n. This means that the algorithm correctly calculates the maximum possible score for an array of size n.

**Inductive Step**: The algorithm always chooses the middle element of the array as the next score, and then recursively calculates the maximum score for the remaining elements. Since the remaining elements are a subset of the original array, and we have assumed that the algorithm works correctly for an array of length n, it follows that the algorithm also works correctly for an array of length n+1

### Running time - Optimal Solution
Denote n as the length of our input array 'arr'.
We will prove our algorithm run in O(n *log(n)):

First note thet we are calling the `max_score_calc` that send our input score and the sorted version of our array into the `max_score` function.
- `max_score(score, sorted(arr))` - The sorted() function has a time complexity of O(n log n)

Afterward for the function `max_score` we have:
- `if len(arr) == 0` : The algorithm first checks if the array is empty. This operation has a constant time complexity of O(1).
- `index = math.ceil(len(arr) / 2) - 1` : Selects the index of middle element of the array using len and ceil takes O(1).
- `next_score = arr[index]` : Selecting an element from a sorted array has a time complexity of O(1).  
- `max_score_calc(next_score, [x + (score - next_score) for x in arr if x != next_score])`: Now we call our function in a recursive way with len(arr) = N-1. Also note that we change each element in the new array this is done by going over all elements in the array which takes O(n) and in the next recursive call this will take O(n- 1), since the length of the new input array decrease.  
Therfore the overall running time of `max_score` is O(n + (n-1) + (n-2) + ... + 1) < O( c * n) = O(n).

So we get our algorithm has running time of O(n log(n) + n) = O(n log(n)).  

This implementation is the optimal one since in order to solve the question me must sort the given array which takes at least O(n log(n)).

### AI third implementation

The following code is the implementation suggested by chatGPT.
The running time of this code is explained as follow:

In this code, S is the initial score and exams is a list of exam marks. The function max_score returns the maximum possible score. The time complexity of this solution is O(N^2), where N is the number of exams. This is because for each exam, we are iterating over all the previous exams to calculate the maximum possible score. The space complexity is O(N), which is used to store the dynamic programming table dp.


In [26]:
def max_score_ai(S, exams):
    exams.sort(reverse=True)
    dp = [0]*len(exams)
    for i in range(len(exams)):
        for j in range(i):
            if exams[i] > exams[j]:
                dp[i] = max(dp[i], dp[j] + exams[i] - exams[j])
            else:
                dp[i] = max(dp[i], dp[j] + S - exams[i])
        dp[i] = max(dp[i], exams[i])
    return dp[-1]

In [28]:
score = int(input("Enter initial personal score: "))
arr = list(map(int, input("Enter exams marks: ").split()))
max_score_ai(score, arr)

Enter initial personal score: 30
Enter exams marks:  13 27 41 59 28 33 39 19 52 48 55 79


112

Clearly the algorithm is wrong since the Max_score for the input given is 105 and not 112.